horovod/run/elastic/driver.py

```diff
@@ -0,0 +1,283 @@
+
+        self._worker_registry = WorkerStateRegistry(self, self._host_manager)
+        self._results = Results()
+        self._shutdown = threading.Event()
+
+        self._discovery_thread = threading.Thread(target=self._discover_hosts)
+        self._discovery_thread.daemon = True
+        self._discovery_thread.start()
+
+    def start(self, np, create_worker_fn):
+        self._create_worker_fn = create_worker_fn
+        self._activate_hosts(np)
+
+    def resume(self):
+        self._activate_hosts(self._min_np)
+
+    def stop(self):
+        self._shutdown.set()
+        self._discovery_thread.join()
+
+    def finished(self):
+        return self._shutdown.is_set()
+
+    def get_results(self):
+        return self._results.get_results()
+
+    def register_worker_server(self, host, slot, addresses, secret_key):
+        self._worker_clients[(host, slot)] = WorkerNotificationClient(
+            addresses, secret_key, self._verbose)
+
+    def record_ready(self, host, slot):
+        self._worker_registry.record_ready(host, slot)
+
+    def world_size(self):
+        return self._world_size
+
+    def local_size(self, host):
+        return len(self._host_assignments[host])
+
+    def get_slot_info(self, host, slot):
+        return self._host_assignments[host][slot] if self.has_rank_assignment(host, slot) \
+            else hosts.INVALID_SLOT_INFO
+
+    def has_rank_assignment(self, host, slot):
+        if self._host_manager.is_blacklisted(host):
+            return False
+        return host in self._host_assignments and len(self._host_assignments[host]) > slot
+
+    def wait_for_available_hosts(self, min_np, min_hosts=1):
+        extra_message = ' An auto-scaling job (`--max-np` > `--num-proc`) also requires that at least two hosts ' \
+                        'are available to resolve compatible network interfaces. If you know which interfaces ' \
+                        'are compatible in your network, set `--nic` to skip this check.' if min_hosts > 1 else ''
+
+        tmout = timeout.Timeout(
+            self._start_timeout,
+            message='Timed out waiting for {{activity}}. Please check that you have '
+                    'enough resources to run at least {min_np} Horovod processes.{extra_message}'
+                    .format(min_np=min_np, extra_message=extra_message))
+
+        self._wait_hosts_cond.acquire()
+        try:
+            while True:
+                current_hosts = self._host_manager.current_hosts
+                if current_hosts.count_available_slots() >= min_np and len(current_hosts.available_hosts) >= min_hosts:
+                    return current_hosts
+                self._wait_hosts_cond.wait(tmout.remaining())
+                tmout.check_time_out_for('minimum number of hosts to become available')
+        finally:
+            self._wait_hosts_cond.release()
+
+    def _activate_hosts(self, min_np):
+        logging.info('wait for available hosts: {}'.format(min_np))
```

Comment:
Actually, we are not waiting for `min_np` hosts here but slots. There is a place now where we wait for `min_hosts`. This naming / wording here can be misleading, maybe `workers` is better than `hosts` here as we use that term in other places already

Reply:
Changed the name to `_activate_workers` and `wait_for_available_slots`.